src/common/stack.rs

```diff
@@ -1,88 +1,419 @@
 use alloc::prelude::*;
 
-#[cfg(feature = "std")]
-use std::error;
-use core::fmt;
+#[derive(Copy, Clone, Debug)]
+pub struct StackOverflow;
 
+/// Pre-allocated, growable stack with upper bound on size.
+///
+/// StackWithLimit is guaranteed never to grow larger than a set limit.
+/// When limit is reached attempting to push to the stack will return
+/// `Err(StackOverflow)`.
+///
+/// Both limit and initial stack size are configurable.
+/// `StackWithLimit` will start out with initial size, but grow when necessary.
 #[derive(Debug)]
-pub struct Error(String);
+pub struct StackWithLimit<T> {
+	stack: Vec<T>,
+	limit: usize,
+}
+
+impl<T> StackWithLimit<T> {
+	/// Create a StackWithLimit with `limit` max size and `initial_size` of pre-allocated
+	/// memory.
+	///
+	/// ```
+	/// # extern crate wasmi;
+	/// # use wasmi::{StackWithLimit, StackSize, FuncRef};
+	/// StackWithLimit::<FuncRef>::new(
+	/// 	StackSize::from_element_count(1024).into_initial(),
+	/// 	StackSize::from_element_count(2048).into_limit(),
+	/// );
+	/// ```
+	///
+	/// Unlimited
+	///
+	/// ```
+	/// # extern crate wasmi;
+	/// # use wasmi::{StackWithLimit, StackSize, RuntimeValue};
+	/// StackWithLimit::<RuntimeValue>::new(
+	/// 	StackSize::from_element_count(1024).into_initial(),
+	/// 	StackSize::unlimited(),
+	/// );
+	/// ```
+	///
+	/// # Errors
+	///
+	/// In debug mode, panics if `initial_size` is larger than `limit`.
+	pub fn new(initial_size: StackSizeInitial<T>, limit: StackSizeLimit<T>) -> StackWithLimit<T> {
+		let initial_size_elements = initial_size.0.element_count();
+		let limit_elements = limit.0.element_count();
+		debug_assert!(
+			limit_elements >= initial_size_elements,
+			"initial_size should not be larger than StackWithLimit limit"
+		);
+		StackWithLimit {
+			stack: Vec::with_capacity(initial_size_elements.min(limit_elements)),
+			limit: limit_elements,
+		}
+	}
 
-impl fmt::Display for Error {
-	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(f, "{}", self.0)
+	/// Create an new StackWithLimit with `limit` max size and `limit` elements pre-allocated
+	///
+	/// ```
+	/// # extern crate wasmi;
+	/// # use wasmi::{StackWithLimit, StackSize};
+	/// let bstack = StackWithLimit::<i32>::with_size(StackSize::from_element_count(2));
+	/// ```
+	pub fn with_size(size: StackSize<T>) -> StackWithLimit<T> {
+		StackWithLimit::new(StackSizeInitial(size), StackSizeLimit(size))
+	}
+
+	/// Attempt to push value onto stack.
+	///
+	/// # Errors
+	///
+	/// Returns Err(StackOverflow) if stack is already full.
+	#[inline]
+	pub(crate) fn push(&mut self, value: T) -> Result<(), StackOverflow> {
+		if self.stack.len() == self.stack.capacity() {
+			if self.stack.len() == self.limit {
+				return Err(StackOverflow);
+			}
+			// grows exponentially, just like Vec normally does
+			// https://doc.rust-lang.org/1.26.0/src/alloc/raw_vec.rs.html#462
+			let desired_len = self
+				.stack
+				.len()
+				.checked_mul(2)
+				.unwrap_or(usize::MAX)
```

Comment:
You can use `saturating_mul` instead of `checked_mul(...).unwrap_or(usize::MAX)`

Reply:
Handy!

Changed it to `saturating_mul`.